pkg/kubectl/cmd/get/get.go

```diff
@@ -544,11 +544,10 @@ func (o *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []string) e
 			w.Flush()
 			w.SetRememberedWidths(nil)
 
-			// TODO: this doesn't belong here
 			// add linebreak between resource groups (if there is more than one)
 			// skip linebreak above first resource group
-			if lastMapping != nil && !o.NoHeaders {
-				fmt.Fprintln(o.ErrOut)
+			if lastMapping != nil && !o.NoHeaders && mapping != nil {
```

Comment:
why was the `&& mapping != nil` added?

Reply:
(dropping it still passes all tests)

Reply:
@liggitt I added this to fix [this reported problem](https://github.com/kubernetes/kubectl/issues/582#issuecomment-506036019). On a second look I think this won't fix the problem because this was originally printing to stderr, not stdout which was where the problem was observed

Reply:
ah. we need a test that demonstrates the issue. the following change reproduces it:

```diff
diff --git a/pkg/kubectl/cmd/get/get_test.go b/pkg/kubectl/cmd/get/get_test.go
index 2e8553a1f0f..7da4844ad60 100644
--- a/pkg/kubectl/cmd/get/get_test.go
+++ b/pkg/kubectl/cmd/get/get_test.go
@@ -395,23 +395,23 @@ func TestGetMultipleTableResourceTypesShowKinds(t *testing.T) {
                        case p == "/namespaces/test/pods" && m == "GET":
                                return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: podTableObjBody(codec, pods.Items...)}, nil
                        case p == "/namespaces/test/replicationcontrollers" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &corev1.ReplicationControllerList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/services" && m == "GET":
                                return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: serviceTableObjBody(codec, svcs.Items...)}, nil
                        case p == "/namespaces/test/statefulsets" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &appsv1.StatefulSetList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/horizontalpodautoscalers" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &autoscalingv1.HorizontalPodAutoscalerList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/jobs" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &batchv1.JobList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/cronjobs" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &batchv1beta1.CronJobList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/daemonsets" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &appsv1.DaemonSetList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/deployments" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &extensionsv1beta1.DeploymentList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
                        case p == "/namespaces/test/replicasets" && m == "GET":
-                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: cmdtesting.ObjBody(codec, &extensionsv1beta1.ReplicaSetList{})}, nil
+                               return &http.Response{StatusCode: 200, Header: cmdtesting.DefaultHeader(), Body: emptyTableObjBody(codec)}, nil
 
                        default:
                                t.Fatalf("request url: %#v,and request: %#v", req.URL, req)
@@ -2581,6 +2581,14 @@ var podColumns = []metav1.TableColumnDefinition{
        {Name: "Readiness Gates", Type: "string", Format: "", Priority: 1},
 }
 
+// build an empty table response
+func emptyTableObjBody(codec runtime.Codec) io.ReadCloser {
+       table := &metav1.Table{
+               ColumnDefinitions: podColumns,
+       }
+       return cmdtesting.ObjBody(codec, table)
+}
+
```

Reply:
the issue is that empty Table responses from the server are not recognized as empty until we are down inside the PrintObj method. Checking `mapping != nil` isn't the right condition to check here. It's not beautiful, but we have the trackingWriter paying attention to whether we have output content... we could use that as a condition to determine whether we should output a blank line, like this:
```diff
diff --git a/pkg/kubectl/cmd/get/get.go b/pkg/kubectl/cmd/get/get.go
index 21114d47186..4b3e90dd167 100644
--- a/pkg/kubectl/cmd/get/get.go
+++ b/pkg/kubectl/cmd/get/get.go
@@ -521,6 +521,8 @@ func (o *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []string) e

        // track if we write any output
        trackingWriter := &trackingWriterWrapper{Delegate: o.Out}
+       // remember how much we've written
+       written := 0

        w := utilprinters.GetNewTabWriter(trackingWriter)
        for ix := range objs {
@@ -546,8 +548,12 @@ func (o *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []string) e

                        // add linebreak between resource groups (if there is more than one)
                        // skip linebreak above first resource group
-                       if lastMapping != nil && !o.NoHeaders && mapping != nil {
-                               fmt.Fprintln(w)
+                       if lastMapping != nil && !o.NoHeaders {
+                               // If we've written output since the last time we started a new set of headers, write an empty line to separate object types
+                               if written != trackingWriter.Written {
+                                       fmt.Fprintln(w)
+                                       written = trackingWriter.Written
+                               }
                        }

                        printer, err = o.ToPrinter(mapping, nil, printWithNamespace, printWithKind)
```

Reply:
That makes sense. Thanks @liggitt ! I will update the PR